ClassFanOutComplexity: false positive on var

Taken from https://github.com/checkstyle/checkstyle/issues/6316#issuecomment-451522961

````
$ cat TestClass.java
public class TestClass {
    void method() {
var x = 1;
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="ClassFanOutComplexity">
  <property name="max" value="0"/>
</module>
    </module>
</module>

$ java -jar checkstyle-8.15-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:1:1: Class Fan-Out Complexity is 1 (max allowed is 0). [ClassFanOutComplexity]
Audit done.
Checkstyle ends with 1 errors.
````

`var` should not be affecting ClassFanOutComplexity by default.
With `var` changed to `Object`, no violation is produced.

https://checkstyle.org/config_metrics.html#ClassFanOutComplexity should ignore var,int type should be counted ... as it is not a class..

@romani Please review.

> This check should ignore var completely?

My initial response is yes.

Based on the JLS,
https://docs.oracle.com/javase/specs/jls/se10/html/jls-14.html#jls-14.4.1
>  If LocalVariableType is var, then let T be the type of the initializer expression when treated as if it did not appear in an assignment context, and were thus a standalone expression (§15.2). The type of the local variable is the upward projection of T with respect to all synthetic type variables mentioned by T (§4.10.5). 

I would say it should behave the same as `T` as if it was a template for a method/class.
> `<T> void method(T param) {}`

But I checked this out with the check, and it flags template classes too.
@romani Do you agree that template classes shouldn't be flagged too?

````
$ cat TestClass.java
public class TestClass {
    <T> void method(T param) {
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="ClassFanOutComplexity">
  <property name="max" value="0"/>
</module>
    </module>
</module>

$ java -jar checkstyle-8.20-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:1:1: Class Fan-Out Complexity is 1 (max allowed is 0). [ClassFanOutComplexity]
Audit done.
Checkstyle ends with 1 errors.
````

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java
@@ -55,6 +55,8 @@ public abstract class AbstractClassCouplingCheck extends AbstractCheck {
     /** Class names to ignore. */
     private static final Set<String> DEFAULT_EXCLUDED_CLASSES = Collections.unmodifiableSet(
         Arrays.stream(new String[] {
+            // reserved type name
+            "var",
             // primitives
             "boolean", "byte", "char", "double", "float", "int",
             "long", "short", "void",

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassDataAbstractionCouplingCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassDataAbstractionCouplingCheck.java
@@ -94,7 +94,7 @@ import com.puppycrawl.tools.checkstyle.api.TokenTypes;
  * Object, Override, Queue, RuntimeException, SafeVarargs, SecurityException, Set, Short,
  * SortedMap, SortedSet, String, StringBuffer, StringBuilder, SuppressWarnings, Throwable,
  * TreeMap, TreeSet, UnsupportedOperationException, Void, boolean, byte, char, double,
- * float, int, long, short, void}.
+ * float, int, long, short, var, void}.
  * </li>
  * <li>
  * Property {@code excludeClassesRegexps} - Specify user-configured regular

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheck.java
@@ -61,7 +61,7 @@ import com.puppycrawl.tools.checkstyle.api.TokenTypes;
  * Object, Override, Queue, RuntimeException, SafeVarargs, SecurityException, Set, Short,
  * SortedMap, SortedSet, String, StringBuffer, StringBuilder, SuppressWarnings, Throwable,
  * TreeMap, TreeSet, UnsupportedOperationException, Void, boolean, byte, char, double,
- * float, int, long, short, void}.
+ * float, int, long, short, var, void}.
  * </li>
  * <li>
  * Property {@code excludeClassesRegexps} - Specify user-configured regular
@@ -96,6 +96,10 @@ import com.puppycrawl.tools.checkstyle.api.TokenTypes;
  *   Date date = new Date(); // Counted, 1
  *   Time time = new Time(); // Counted, 2
  *   Place place = new Place(); // Counted, 3
+ *   int value = 10; // int is ignored due to default excludedClasses property
+ *   void method() {
+ *     var result = "result"; // var is ignored due to default excludedClasses property
+ *   }
  * }
  * </pre>
  * <p>

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java
@@ -302,6 +302,16 @@ public class ClassFanOutComplexityCheckTest extends AbstractModuleTestSupport {
             getNonCompilablePath("InputClassFanOutComplexityRecords.java"), expected);
     }
 
+    @Test
+    public void testClassFanOutComplexityIgnoreVar() throws Exception {
+        final DefaultConfiguration checkConfig =
+            createModuleConfig(ClassFanOutComplexityCheck.class);
+        checkConfig.addAttribute("max", "0");
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verify(checkConfig,
+            getNonCompilablePath("InputClassFanOutComplexityVar.java"), expected);
+    }
+
     /**
      * We cannot reproduce situation when visitToken is called and leaveToken is not.
      * So, we have to use reflection to be sure that even in such situation

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityVar.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityVar.java
@@ -1,0 +1,19 @@
+//non-compiled with javac: Compilable with Java10
+package com.puppycrawl.tools.checkstyle.checks.metrics.classfanoutcomplexity;
+
+import javax.naming.NamingException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/* Config:
+ * max  = 0
+ */
+class InputClassFanOutComplexityVar { // ok
+    void method() {
+        var x = 1;
+    }
+}

--- a/src/xdocs/config_metrics.xml
+++ b/src/xdocs/config_metrics.xml
@@ -310,7 +310,7 @@ public class Test
                 SafeVarargs, SecurityException, Set, Short, SortedMap, SortedSet,
                 String, StringBuffer, StringBuilder, SuppressWarnings, Throwable,
                 TreeMap, TreeSet, UnsupportedOperationException, Void, boolean,
-                byte, char, double, float, int, long, short, void</code>
+                byte, char, double, float, int, long, short, var, void</code>
               </td>
               <td>5.7</td>
             </tr>
@@ -693,7 +693,7 @@ class Foo {
                 SafeVarargs, SecurityException, Set, Short, SortedMap, SortedSet,
                 String, StringBuffer, StringBuilder, SuppressWarnings, Throwable,
                 TreeMap, TreeSet, UnsupportedOperationException, Void, boolean,
-                byte, char, double, float, int, long, short, void</code>
+                byte, char, double, float, int, long, short, var, void</code>
               </td>
               <td>5.7</td>
             </tr>
@@ -739,6 +739,10 @@ class InputClassComplexity {
   Date date = new Date(); // Counted, 1
   Time time = new Time(); // Counted, 2
   Place place = new Place(); // Counted, 3
+  int value = 10; // int is ignored due to default excludedClasses property
+  void method() {
+    var result = "result"; // var is ignored due to default excludedClasses property
+  }
 }
         </source>
 

